cli/src/commands/sign.rs

```diff
@@ -0,0 +1,161 @@
+    #[arg(long)]
+    allow_not_mine: bool,
+}
+
+pub fn check_commits_authored_by<'a>(
+    commits: impl IntoIterator<Item = &'a Commit>,
+    user_email: &str,
+    hint: &str,
+) -> Result<(), CommandError> {
+    let bad_commits = commits
+        .into_iter()
+        .filter(|commit| commit.author().email != user_email)
+        .collect_vec();
+
+    match bad_commits.len() {
+        0 => Ok(()),
+        1 => Err(user_error_with_hint(
+            format!(
+                "Commit {} is not authored by you.",
+                short_commit_hash(bad_commits[0].id()),
+            ),
+            hint,
+        )),
+        2.. => Err(user_error_with_hint(
+            format!(
+                "The following commits are not authored by you:\n  {}",
+                bad_commits
+                    .into_iter()
+                    .ids()
+                    .map(short_commit_hash)
+                    .join("\n  ")
+            ),
+            hint,
+        )),
+    }
+}
+
+pub fn cmd_sign(ui: &mut Ui, command: &CommandHelper, args: &SignArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    let commits: IndexSet<Commit> = workspace_command
+        .parse_union_revsets(ui, &args.revisions)?
+        .evaluate_to_commits()?
+        .try_collect()?;
+
+    workspace_command.check_rewritable(commits.iter().ids())?;
+
+    if !args.allow_not_mine {
+        check_commits_authored_by(
+            &commits,
+            command.settings().user_email(),
+            "Use --allow-not-mine to sign anyway.",
+        )?;
+    }
+
+    let mut tx = workspace_command.start_transaction();
+
+    let behavior = if args.allow_not_mine {
+        SignBehavior::Force
+    } else {
+        SignBehavior::Own
+    };
+    let mut signed_commits = vec![];
+    tx.repo_mut().transform_descendants(
+        commits.iter().ids().cloned().collect_vec(),
+        |rewriter| {
+            if rewriter.old_commit().is_signed() {
+                // Don't resign commits, which are already signed by me. For people using
+                // hardware devices for signatures, resigning commits is
+                // cumbersome.
+                return Ok(());
+            }
+
+            if commits.contains(rewriter.old_commit()) {
+                let commit_builder = rewriter.reparent();
+                let new_commit = commit_builder
+                    .set_sign_key(args.key.clone())
+                    .set_sign_behavior(behavior)
+                    .write()?;
+                signed_commits.push(new_commit);
+            }
+            Ok(())
+        },
+    )?;
+
+    if let Some(mut formatter) = ui.status_formatter() {
+        match signed_commits.len() {
+            0 => (),
+            1 => {
+                write!(formatter, "Signed commit ")?;
+                tx.base_workspace_helper()
+                    .write_commit_summary(formatter.as_mut(), &signed_commits[0])?;
```

Comment:
use `tx.write_commit_summary()` and `tx.commit_summary_template()` because `signed_commits[N]` is rewritten commits. (also applies to unsign)